src/lib/operators/join_index.cpp

```diff
@@ -0,0 +1,358 @@
+                                     _scan_type);
+}
+
+std::shared_ptr<const Table> JoinIndex::_on_execute() {
+  _create_table_structure();
+
+  _perform_join();
+
+  return _output_table;
+}
+
+void JoinIndex::_create_table_structure() {
+  _output_table = std::make_shared<Table>();
+
+  _left_in_table = _input_left->get_output();
+  _right_in_table = _input_right->get_output();
+
+  _left_column_id = _column_ids.first;
+  _right_column_id = _column_ids.second;
+
+  const bool left_may_produce_null = (_mode == JoinMode::Right || _mode == JoinMode::Outer);
+  const bool right_may_produce_null = (_mode == JoinMode::Left || _mode == JoinMode::Outer);
+
+  // Preparing output table by adding columns from left table
+  for (ColumnID column_id{0}; column_id < _left_in_table->column_count(); ++column_id) {
+    auto nullable = (left_may_produce_null || _left_in_table->column_is_nullable(column_id));
+    _output_table->add_column_definition(_left_in_table->column_name(column_id), _left_in_table->column_type(column_id),
+                                         nullable);
+  }
+
+  // Preparing output table by adding columns from right table
+  for (ColumnID column_id{0}; column_id < _right_in_table->column_count(); ++column_id) {
+    auto nullable = (right_may_produce_null || _right_in_table->column_is_nullable(column_id));
+    _output_table->add_column_definition(_right_in_table->column_name(column_id),
+                                         _right_in_table->column_type(column_id), nullable);
+  }
+}
+
+// inner join loop that joins two columns via their iterators
+template <typename BinaryFunctor, typename LeftIterator, typename RightIterator>
+void JoinIndex::_join_two_columns(const BinaryFunctor& func, LeftIterator left_it, LeftIterator left_end,
+                                  RightIterator right_begin, RightIterator right_end, const ChunkID chunk_id_left,
+                                  const ChunkID chunk_id_right, std::vector<bool>& left_matches) {
+  const auto to_row_id = [chunk_id_right](ChunkOffset chunk_offset) { return RowID{chunk_id_right, chunk_offset}; };
+
+  // new inner loop
+  const auto& r_chunk = _right_in_table->get_chunk(chunk_id_right);
+  const auto indices = r_chunk.get_indices(std::vector<ColumnID>{_right_column_id});
+
+  Assert(indices.size() >= 1, "No index found for the right column");
+
+  // We assume the first index to be good. Evaluating index performance is not task of the join.
+  const auto index = indices.front();
+
+  if (index != nullptr && !_fallback) {
+    // We have a proper index and can use it, so we do an index join for this chunk combination
+    // outer loop, we keep this
+    for (; left_it != left_end; ++left_it) {
+      const auto left_value = *left_it;
+      if (left_value.is_null()) continue;
+
+      auto var = AllTypeVariant(left_value.value());
+
+      auto comp_values = std::vector<AllTypeVariant>();
+
+      comp_values.push_back(var);
```

Comment:
We should think about extending the BaseIndex's interface to include a non-vectorized version.

Should be straight-forward and since we're within a loop here, it might also make sense from a performance perspective.

Reply:
Will add an issue for this.